File: bindings/gumjs/gumjscriptbundle.m
/*
 * Copyright (C) 2015 Ole André Vadla Ravnås <oleavr@nowsecure.com>
 *
 * Licence: wxWindows Library Licence, Version 3.1
 */

#import "gumjscriptbundle.h"

@implementation GumScriptBundle

+ (void)load:(const GumScriptSource *)sources intoContext:(JSContext *)context
{
  const GumScriptSource * cur;

  for (cur = sources; cur->name != NULL; cur++)
  {
    gchar * str = g_strjoinv (NULL, (gchar **) cur->chunks);
    NSString * source = [NSString stringWithUTF8String:str];
    g_free (str);

    NSString * filename = [[NSString stringWithUTF8String:cur->name]
                                  stringByAppendingString:@".js"];
    NSURL * url = [NSURL URLWithString:
        [@"file:///" stringByAppendingString:filename]];

    [context evaluateScript:source
              withSourceURL:url];
    g_assert (context.exception == nil);
  }
}

@end

File: bindings/gumjs/gumjscriptbundle.c
/*
 * Copyright (C) 2015 Ole André Vadla Ravnås <oleavr@nowsecure.com>
 *
 * Licence: wxWindows Library Licence, Version 3.1
 */

#include "gumjscriptbundle.h"

#include "gumjscriptcore.h"

void
gum_script_bundle_load (const GumScriptSource * sources,
                        JSContextRef ctx)
{
  JSObjectRef global;
  const GumScriptSource * cur;

  global = JSContextGetGlobalObject (ctx);

  for (cur = sources; cur->name != NULL; cur++)
  {
    JSStringRef source, url;
    gchar * str;
    JSValueRef result, exception;

    str = g_strjoinv (NULL, (gchar **) cur->chunks);
    source = JSStringCreateWithUTF8CString (str);
    g_free (str);

    str = g_strconcat ("file:///", cur->name, NULL);
    url = JSStringCreateWithUTF8CString (str);
    g_free (str);

    result = JSEvaluateScript (ctx, source, global, url, 1, &exception);
    if (result == NULL)
      _gum_script_panic (exception, ctx);

    JSStringRelease (url);
    JSStringRelease (source);
  }
}
